Qpf/Qpf/Multivariate/Constructions/DeepThunk.lean

```diff
@@ -0,0 +1,169 @@
+import Mathlib.Data.QPF.Multivariate.Constructions.Comp
+import Mathlib.Data.QPF.Multivariate.Constructions.Prj
+import Mathlib.Data.QPF.Multivariate.Constructions.Cofix
+import Qpf.Macro.Comp
+
+namespace MvQPF
+
+/-!
+  # DeepThunk
+
+  `DeepThunk` corresponds to the (co-)trace of a given polynomial functor.
+  These allows the original pfunctor to inject into them allowing for the most general state of a co-recursive principle.
+
+  ## Stronger generaliziation
+
+  Currently the functors are mapped as follows `F := cofix f α in α` to `DT := cofix f (β ⊕ α) in α`.
+  This could be generalized to `DT := cofix f ( (β ⊕ α) × (F → F) )` which would allow for modifying the returned value of a deeper call.
+  Note this is still a QPF as (F → F) is constant, also note the function could not be (F → DT) as this would be unproductive
+  (think of the function that simply skips the first value then yields a recursive construct).
+  This generaliziation would be useful in expression streams of for example the natural numbers as:
+
+  codef nats : Colist ℕ := .cons 0 (nats.map Nat.succ)
+
+  which would be productive though strange. This generalizes to nested guarded recursive structures.
+  Unproductive unguarded structures are limited in the same way they currently are.
+-/
+
+/--
+  This is a fresh sum-type that represents the actions that can be taken within a `DeepThunk`.
+  These are that a user can either
+
+  1. recall - hand responsibility back to the co-recursor
+  2. cont   - continue constructing a DeepThunk
+
+  These we're previously represented as a sum type but  
+-/
+inductive DTSum (α β : Type u) : Type u
+  | recall (v : α)
+  | cont   (v : β)
+
+namespace DTSum
+abbrev F := @TypeFun.ofCurried 2 DTSum
+
+def box {Γ : TypeVec 2} : F Γ → QpfSum' Γ
+  | recall a => MvQPF.Sum.inl a
+  | cont   b => MvQPF.Sum.inr b
+
+def unbox {Γ : TypeVec 2} : QpfSum' Γ → F Γ
+  | ⟨i, f⟩ => match i with
+    | .fz   => recall (f 1 .fz)
+    | .fs 0 => cont   (f 0 .fz)
+
+def equiv {Γ} : F Γ ≃ QpfSum' Γ :=
+{
+  toFun   := box
+  invFun  := unbox
+  left_inv  := by intro x; cases x <;> rfl
+  right_inv := by
+    intro x
+    rcases x with ⟨(i : PFin2 _), f⟩
+    dsimp only [box, unbox, Sum.inl, Sum.inr]
+    fin_cases i <;> {
+      simp only [Function.Embedding.coeFn_mk, PFin2.ofFin2_fs, PFin2.ofFin2_fz,
+        Fin2.instOfNatFin2HAddNatInstHAddInstAddNatOfNat, Nat.rec_zero]
+      apply congrArg
+      funext i; fin_cases i <;> (funext (j : PFin2 _); fin_cases j)
+      rfl
+    }
+}
+
+
+end DTSum
```

Comment:
It might be cleaner to build an equivalence with regular `Sum`, which should have a QPF instance as well.

If you want to be extra spicy: what happens if we define `DTSum` using `data`?
That should give you the QPF instance for free